### ISSUE TITLE
Generic wildcard types should not be used in return parameters

### ISSUE THREAD

**Issue by ilentt** [Neloop/jaclp#3]
``` 
    src/main/java/cz/polankam/security/acl/IPermissionsService.java
    /**
     * For the given textual representation of resource return its resource
     * repository.
     * Should throw in case of not defined resource.
     * @param resource textual resource representation
     * @return repository from which resource object can be acquired
     */
    IResourceRepository<?, ?> getResource(String resource);
```
SonarLint recomment that should not used Generic wildcard types in return parameters. In [Effective Java Item 28](https://stackoverflow.com/questions/22815023/generic-wildcard-types-should-not-be-used-in-return-parameters) also comment that **If the user of a class has to think about wildcard types, there is probably something wrong with the class’s API.** So do we have any alternative solution to fit clean code recommendation?

**Comment by Neloop** [comment]
Let me look into it and see if it can be done cleaner...

**Comment by Neloop** [comment]
Okay, I investigated a bit and there are 3 ways which came to my mind how to change the API, unfortunatelly none of it is perfect

1. Return `IResourceRepository` without parametrized types, this of course spawns different SonarLint warning about 'Raw use of parametrized types', so it seems very similar to the current situation.

```{java}
IResourceRepository getResource(String resource);
```

2. Return `IResourceRepository<Object, Object>`, from the `IPermissionsService` interface point of view this is fine and without warnings. The problem is in implementation of `IPermissionsService` where you have to either use raw parametrized types or perform unchecked assignments.

```{java}
// IPermissionsService.java
IResourceRepository<Object, Object> getResource(String resource);

// DemoPermissionsService.java
private final Map<String, IResourceRepository<Object, Object>> resources = new HashMap<>();
// ...
resources.put("group", (IResourceRepository) new DemoGroupRepository());
```

3. Do not use generics in `IResourceRepository`, this breaks my original intention to have `findById` method signature the same as in `CrudRepository` and the signatures would clash. Therefore in this case your JPA repositories cannot implement `IResourceRepository` interface and it have to be handled differently, most probably directly in the `IPermissionsService` implementation

```{java}
// IResourceRepository.java
public interface IResourceRepository {
    Optional<Object> findById(Object id);
}

// DemoPermissionsService.java
private final Map<String, IResourceRepository<Object, Object>> resources = new HashMap<>();
// ...
resources.put("user", (id) -> Optional.ofNullable(userRepository.findById((Long) id).orElse(null)));
resources.put("group", (id) -> Optional.ofNullable(groupRepository.findById((Long) id).orElse(null)));
```

By looking at those solutions, I have to say that I am leaning towards solution number 3. From the warnings perspective it is the cleanest and considering that I work with resources without knowing its type, having Object there seems right. Any other suggestions, ideas?

**Comment by Neloop** [comment]
I took the third approach and removed generics from `IResourceRepository`. Implemented in here 4389d48a034bf6f8d93a19970d21b7ec809702ac

It is part of `3.0` release, which is available in Maven Central.
